Memory usage optimizations

andreakarasho opened the pull request:
- Implemented `System.Buffers` & `System.Memory` [and relative subdeps] for net472.
- High use of `ArrayPool<T>` to rent temp buffers avoiding GC time wast and allocation time
- Added new struct `ValueStringBuffer` to create and manage strings with 0~ allocation [i got the idea from the good @kamronbatman]
- `PixelPicker` class to store compressed non zero rgb pixels instead of allocate `UOTexture::Data` field [from UltimaXNA]
- Added a `UnmanagedMemoryPool` struct that stores a linkedlist of unmanaged data. It works like `ArrayPool<T>` not sure about the usage but it's cool.

NOTE: `System.Buffers.ArrayPool<T>.Shared.Return` accept as 2nd param `clearArray`. I set it to `true` when creating `uint[]` otherwhise textures will contains garbage.

andreakarasho committed:
api update

andreakarasho committed:
experimental C# 9.0 to support function pointers

andreakarasho committed:
other delegates

andreakarasho committed:
space

andreakarasho committed:
Merge branch 'dev' into plugin-api

andreakarasho committed:
marshal string

andreakarasho committed:
ptr for string

andreakarasho committed:
MONO_THREADS_SUSPEND=preemptive

andreakarasho committed:
dllmap

andreakarasho committed:
encode heap

andreakarasho committed:
Merge branch 'dev' into plugin-api

andreakarasho committed:
Merge branch 'dev' into plugin-api

andreakarasho committed:
Merge branch 'dev' into plugin-api

andreakarasho committed:
+ System.Buffers
+ System.Memory
+ large use of arraypool to optimize performances
+ ValueStringBuilder [internet and minor from modernuo]
+ pixel picker [from UltimaXNA]
+ reduced GC pressure about 40%

andreakarasho committed:
+ removed "Data" fileds from UOTexture

andreakarasho committed:
+ dlls

andreakarasho committed:
+ fix pos cliloc translation

andreakarasho committed:
+ again pool

andreakarasho committed:
+ plugin fix

andreakarasho committed:
+ added support for old cuoapi for retrocompatibility

andreakarasho committed:
Merge branch 'dev' into plugin-api

andreakarasho committed:
+ fix asm load

andreakarasho committed:
merge

andreakarasho committed:
merge

andreakarasho committed:
Merge branch 'dev' into memory-optimizations

andreakarasho committed:
merge

andreakarasho committed:
merge

andreakarasho committed:
removed entiticollection, replaced with dictionaries

andreakarasho committed:
+ improved stackdatareader

andreakarasho committed:
+ stackdatawriter
+ few packets ported to the new stackdatawriter

andreakarasho committed:
minor

andreakarasho committed:
+ packets

andreakarasho committed:
+ packets fill zeroing

andreakarasho committed:
minor

andreakarasho committed:
zeroing

andreakarasho committed:
fix containbs

andreakarasho committed:
fix contains 2

andreakarasho committed:
fix house customization coords

andreakarasho committed:
Merge branch 'dev' into memory-optimizations

andreakarasho committed:
+ revert back to c#7.3

andreakarasho committed:
bodytext

andreakarasho committed:
the size of the array is w * h

andreakarasho committed:
+ fix highlight for ascii chars

andreakarasho committed:
+ improved worldmap

andreakarasho committed:
+ check for disposed wmap

kamronbatman commented on src/IO/DataReader.cs:
```diff
@@ -226,19 +227,24 @@ internal string ReadASCII(int size)
-
-            for (int i = 0; i < size; i++)
+            ValueStringBuilder sb = new ValueStringBuilder(size);
             {
```
Is this a naked context? I think this is generally not recommended unless you have a specific need for it since this can create another stack. In this particular case it doesn't look needed.

andreakarasho replied:
done! ty

kamronbatman commented on src/IO/Resources/ClilocLoader.cs:
```diff
@@ -197,157 +217,165 @@ public string Translate(int clilocNum, string arg = "", bool capitalize = false)
 
-            arguments.Add(arg);
+            // store index locations
+            Point* locations = stackalloc Point[++totalArgs];
```
You brought in `Span/Memory`, so I think this can be changed to `Span<Point>` and the unsafe dropped?

kamronbatman commented on src/IO/Resources/FontsLoader.cs:
```diff
@@ -543,66 +546,71 @@ ushort flags
+            int strLen = str.Length;
 
-            if (IsUsingHTML)
+            ValueStringBuilder sb = new ValueStringBuilder(strLen);
```
If we can predetermine the end size of the string, we can avoid resizes, which means we generally don't need rented buffers unless the Span<char> will be large.
There is a break-even point where stackallocating a buffer is slower than calling out to the array pool. This is because for right now, stack allocating always clears the the buffer. IIRC, it's like 100KB or so which is usually plenty.
So if needed, you can do an explicit check for the size of the final string (if it is known), and either stackalloc, or rent accordingly.

andreakarasho replied:
makes sense. I'll check it out

kamronbatman commented on src/IO/StackDataReader.cs:
```diff
@@ -34,77 +47,247 @@ public StackDataReader(byte[] data, long len)
+
+        public string ReadASCII(int size)
+        {
+            ValueStringBuilder sb = new ValueStringBuilder(size);
```
I think it is faster to use `Encoding.ASCII.GetString()` because the conversion is vectorized. We should benchmark to see.
But that might be a .NET Core/5 optimization, not .NET Framework. 🤔

andreakarasho replied:
From my experience i met some server [POL/Sphere] that sends garbage inside the string like:
` 'C', 'L', 'A', 'S', 'S', 'I', 'C', '\0', 'U', 'O' ` but the original client understand it as `CLASSICUO`.
Using `Encoding.ASCII.GetString()` will cut the string at '\0'

kamronbatman replied:
This is true, and that is due to fixed size vs non-fixed size IIRC. If that is the case, you would need to remove all invalid chars, otherwise you can use ASCII get string.

Personally I would not support old/bugged client behavior. It is probably an edge case anyway. ModernUO doesn't support this either because bad behavior is bad...

kamronbatman commented on src/IO/StackDataWriter.cs:
```diff
@@ -0,0 +1,451 @@
+            {
+                char c = str[i];
+
+                if (c != '\0')
```
When I did testing on this, I found that doing an `indexOf` from a `ReadOnlySpan<char>`, slicing from that point, counting the chars and multiplying by 2, then getting an equal sized buffer and just doing `Encoding.Unicode.GetBytes()` to write directly to it was actually faster.

I highly recommend benchmarks for this kind of thing since regular for loops are slower than whatever magic they are doing under the hood (from my experience). This is also specific to .NET Core/5, I am not sure how well .NET Framework performs for encoding.

andreakarasho replied:
https://github.com/andreakarasho/ClassicUO/pull/1370/commits/9ea4049a7c4d4b4c079b305267694c55a001bb88
Done some mod, what do you think about?

kamronbatman replied:
Looks great!

kamronbatman commented on src/Network/Packets.cs:
```diff
@@ -41,1708 +42,4335 @@
-            WriteByte(0xFF);
-        }
-    }
+                writer.WriteUInt8(0x20);
```
Fun fact, if you benchmark this against doing the same thing but with 8byte values, you will find the latter is much faster.

andreakarasho committed:
Merge branch 'dev' into memory-optimizations

andreakarasho committed:
+ removed useless { }

andreakarasho committed:
+ BinaryPrimitives
+ writestring with encoding

andreakarasho committed:
Merge branch 'dev' into memory-optimizations

andreakarasho committed:
Merge branch 'dev' into memory-optimizations

kamronbatman commented on src/IO/StackDataWriter.cs:
```diff
@@ -0,0 +1,344 @@
+
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteZero(int count)
```
It is more efficient to write a byte if it is 1, a ushort if 2, a uint if 4, a ulong if 8, or use `Span<T>.Fill()` if bigger.
I believe Fill uses vectors of all 0's with low level mem copies to keep it efficient.

andreakarasho replied:
net472 does not have tthe `Fill` method :(

kamronbatman replied:
Even with `Span<>` pulled in from nuget? 🤔 
If not then you should do it manually.. and write in chunks of 8bytes and step down accordingly to decrease the amount of looping.
Also I recommend using a directive here:

```cs
#if NET_CORE
// use fill
#else
// Use ghetto replacement
#endif
```

andreakarasho replied:
According to the doc, the `Fill` method is part of the `System.Runtime`    ¯\_(ツ)_/¯

kamronbatman replied:
So you might want to make a `Clear` method and do something like:
if 1 byte, write 0 byte
if 2 bytes, write 0 short
if 4 bytes, write 0 int
if 8+ bytes, write 0 long until less than 8 bytes, then write the remaining amounts.

Something like that I guess.

andreakarasho replied:
Do not ask me what happened but today i can see the `Fill` method
this should be ok now:

```csharp
 [MethodImpl(IMPL_OPTION)]
        public void WriteZero(int count)
        {
            if (count > 0)
            {
                EnsureSize(count);

                _buffer.Slice(Position, count).Fill(0);

                Position += count;
            }
        }
```

andreakarasho committed:
rebase

andreakarasho committed:
Merge branch 'memory-optimizations' of https://github.com/andreakarasho/ClassicUO into memory-optimizations

andreakarasho committed:
use custom font for target indicator range

andreakarasho committed:
span fill zero

andreakarasho committed:
+ write packet length using Fill method

andreakarasho committed:
Merge branch 'dev' into memory-optimizations

andreakarasho committed:
minor improvements in encryption

andreakarasho committed:
+ rename some file
+ deleted old class to write/read packets

andreakarasho committed:
removed packets

andreakarasho committed:
+ aligned plugins & uolive packets

andreakarasho committed:
+ replaced PacketBufferReader with the common StackDataReader

andreakarasho committed:
+ more test for garbage into the string buffer

andreakarasho committed:
minor

andreakarasho committed:
removed comment

andreakarasho committed:
mrege

andreakarasho committed:
+ better string reading for '\0' case. Original client [tested 70910] cuts strings as well when founding '\0' char

andreakarasho committed:
some tests about strings

andreakarasho committed:
+ removed byte[] allocation in stackdatareader and replaced with a fill method
+ replaced Prompt ASCII/Unicode byte[] with an ulong
+ datawriter now supports write64

andreakarasho committed:
+ few stackalloc char

andreakarasho committed:
+ expose a bool to read string safe
+ removed old methods about strings
